### ISSUE TITLE
protocol: KeepAlive responses should be restricted/rate limited.

### ISSUE THREAD

**Issue by Yawning** [ricochet-im/ricochet#281]
I said I'll fork and file a pull request with spec changes that I think are important, but it'll be a while before I can do that, and this is important enough that I should document it sooner rather than later.

This would be both a spec change and a code change.  Currently `KeepAlive` control packets are allowed at any time during the connection lifespan, with no limit on the number of concurrent outstanding keep alives, or rate limit on the responses.

The spec should be changed to include:
 * "Implementations MUST NOT send `KeepAlive` messages before the client has authenticated."
 * "Implementations MUST terminate the connection when a `KeepAlive` request has been received before the client has authenticated."
 * "Implementations MAY ignore `KeepAlive` messages at any time."
 * "Implementations SHOULD limit sending KeepAlive requests and responses to a reasonable value."

The current code will happily accept `KeepAlives` whenever, and will send an unlimited number of responses, even before the peer is authenticated (Though the known purpose timeout somewhat limits the extent of bad) which is suboptimal.

**Comment by special** [comment]
I agree with all of these, although if implementations may ignore `KeepAlive`, is it still useful in any way? We couldn't terminate the connection based on not getting a response.

I bet this is not the only easy denial-of-service-ish opportunity available, although it gets points for being pre-authentication. Good stuff.

**Comment by Yawning** [comment]
> I agree with all of these, although if implementations may ignore KeepAlive, is it still useful in any way? We couldn't terminate the connection based on not getting a response.

It's still useful, TCP Keep-Alives have similar semantics.  What good implementations do there is drop the connection after a sufficient number of keep-alives fail to elicit a response.  I don't see why this couldn't be done with Ricochet.

I'm more concerned that overly aggressive keep-alive behavior can lead to bad, than idle connections taking a while to be re-established, but this is a network heath vs UX tradeoff so more thought will be needed.

As a side note, it would be nice to be able to specify a tag in a KeepAlive request that gets echoed (Eg: a `uint32_t`), so KeepAlives request/response pairs can be used for latency measurements.
